ng update wrote wrong path of main.ts and polyfill.ts in tsconfig.app.json

<!--🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅

Oh hi there! 😄

To expedite issue processing please search open and closed issues before submitting a new one.
Existing issues often contain information about workarounds, resolution, or progress updates.

🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅-->


# 🐞 bug report

### Affected Package
<!-- Can you pin-point one or more @angular/* packages as the source of the bug? -->
<!-- ✍️edit: --> N/A


### Is this a regression?

<!-- Did this behavior use to work in the previous version? -->
<!-- ✍️--> Yes, the previous version in which this bug was not present was: v8


### Description

<!-- ✍️--> After upgrading to v9, I found that path of main.ts and polyfill.ts are wrongly written in tsconfig.app.json which makes the app unusable.


## 🔬 Minimal Reproduction
<!--
Please create and share minimal reproduction of the issue starting with this template: https://stackblitz.com/fork/angular-issue-repro2
-->
<!-- ✍️--> 
Checkout this commit https://github.com/SachinShekhar/angular-ssr-prerendering-bug/commit/593f5999bd356b4ce98ac27db4d24dd797423c43 and run `ng update @angular/core @angular/cli --next`

The result will be https://github.com/SachinShekhar/angular-ssr-prerendering-bug/commit/5528e5fc9baec960a64a4cbb84db72c0561f2df4#diff-d090b1ccd7cff3dfff369e150143bb09

<!--
If StackBlitz is not suitable for reproduction of your issue, please create a minimal GitHub repository with the reproduction of the issue.
A good way to make a minimal reproduction is to create a new app via `ng new repro-app` and add the minimum possible code to show the problem. 
Share the link to the repo below along with step-by-step instructions to reproduce the problem, as well as expected and actual behavior.

Issues that don't have enough info and can't be reproduced will be closed.

You can read more about issue submission guidelines here: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#-submitting-an-issue
-->

## 🔥 Exception or Error
<pre><code>
<!-- If the issue is accompanied by an exception or an error, please share it below: -->
<!-- ✍️-->
ERROR in error TS6053: File 'x:/Bugs/angular-ssr-prerendering-bug/apps/angular-ssr-prerendering-bug/apps/angular-ssr-prerendering-bug/src/main.ts' not found.
error TS6053: File 'x:/Bugs/angular-ssr-prerendering-bug/apps/angular-ssr-prerendering-bug/apps/angular-ssr-prerendering-bug/src/polyfills.ts' not found.
</code></pre>
This error is thrown when I run `ng serve`.

## 🌍  Your Environment

**Angular Version:**
<pre><code>
<!-- run `ng version` and paste output below -->
<!-- ✍️-->
Angular CLI: 9.0.0-rc.11
Node: 12.6.0
OS: win32 x64

Angular: 9.0.0-rc.11
... animations, cli, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... platform-server, router
Ivy Workspace: Yes

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.900.0-rc.11
@angular-devkit/build-angular     0.900.0-rc.11
@angular-devkit/build-optimizer   0.900.0-rc.11
@angular-devkit/build-webpack     0.900.0-rc.11
@angular-devkit/core              9.0.0-rc.11
@angular-devkit/schematics        9.0.0-rc.11
@ngtools/webpack                  9.0.0-rc.11
@nguniversal/builders             9.0.0-rc.1
@nguniversal/common               9.0.0-rc.1
@nguniversal/express-engine       9.0.0-rc.1
@schematics/angular               9.0.0-rc.11
@schematics/update                0.900.0-rc.11
rxjs                              6.5.4
typescript                        3.7.5
webpack                           4.41.2
</code></pre>

**Anything else relevant?**
<!-- ✍️Is this a browser specific issue? If so, please specify the browser and version. -->

<!-- ✍️Do any of these matter: operating system, IDE, package manager, HTTP server, ...? If so, please mention it below. -->
This project uses [NX](https://github.com/nrwl/nx) schematics. I've a feeling that it is the culprit.

--- a/packages/schematics/angular/migrations/update-9/update-app-tsconfigs.ts
+++ b/packages/schematics/angular/migrations/update-9/update-app-tsconfigs.ts
@@ -7,14 +7,20 @@
  */
 import { JsonAstObject, logging } from '@angular-devkit/core';
 import { Rule, Tree, UpdateRecorder } from '@angular-devkit/schematics';
-import { posix } from 'path';
+import { dirname, relative } from 'path';
 import {
   findPropertyInAstObject,
   insertPropertyInAstObjectInOrder,
   removePropertyInAstObject,
 } from '../../utility/json-utils';
 import { Builders } from '../../utility/workspace-models';
-import { getAllOptions, getTargets, getWorkspace, readJsonFileAsAstObject } from './utils';
+import {
+  forwardSlashPath,
+  getAllOptions,
+  getTargets,
+  getWorkspace,
+  readJsonFileAsAstObject,
+} from './utils';
 
 /**
  * Update the tsconfig files for applications
@@ -107,15 +113,18 @@ function updateTsConfig(tree: Tree, builderConfig: JsonAstObject, builderName: B
       const files = findPropertyInAstObject(tsConfigAst, 'files');
       if (!files) {
         const newFiles: string[] = [];
+        const tsConfigDir = dirname(forwardSlashPath(tsConfigPath));
 
         const mainOption = findPropertyInAstObject(option, 'main');
         if (mainOption && mainOption.kind === 'string') {
-          newFiles.push(posix.relative(posix.dirname(tsConfigPath), mainOption.value));
+          newFiles.push(
+            forwardSlashPath(relative(tsConfigDir, forwardSlashPath(mainOption.value))));
         }
 
         const polyfillsOption = findPropertyInAstObject(option, 'polyfills');
         if (polyfillsOption && polyfillsOption.kind === 'string') {
-          newFiles.push(posix.relative(posix.dirname(tsConfigPath), polyfillsOption.value));
+          newFiles.push(
+            forwardSlashPath(relative(tsConfigDir, forwardSlashPath(polyfillsOption.value))));
         }
 
         if (newFiles.length) {

--- a/packages/schematics/angular/migrations/update-9/update-app-tsconfigs_spec.ts
+++ b/packages/schematics/angular/migrations/update-9/update-app-tsconfigs_spec.ts
@@ -8,6 +8,7 @@
 
 import { EmptyTree } from '@angular-devkit/schematics';
 import { SchematicTestRunner, UnitTestTree } from '@angular-devkit/schematics/testing';
+import { getWorkspaceTargets, updateWorkspaceTargets } from './update-workspace-config_spec';
 
 // tslint:disable-next-line: no-any
 function overrideJsonFile(tree: UnitTestTree, path: string, newContent: object) {
@@ -60,6 +61,32 @@ describe('Migration to version 9', () => {
       overrideJsonFile(tree, 'tsconfig.app.json', defaultTsConfigOptions);
       const tree2 = await schematicRunner.runSchematicAsync('workspace-version-9', {}, tree.branch()).toPromise();
       const { exclude, files } = JSON.parse(tree2.readContent('tsconfig.app.json'));
+      expect(exclude).toBeUndefined();
+      expect(files).toEqual(['src/main.ts', 'src/polyfills.ts']);
+    });
+
+    it('should resolve paths correctly even if they are using windows separators', async () => {
+      const tree2 = await schematicRunner
+        .runExternalSchematicAsync(
+          require.resolve('../../collection.json'),
+          'application',
+          {
+            name: 'another-app',
+          },
+          tree,
+        )
+        .toPromise();
+
+      const tsCfgPath = 'projects/another-app/tsconfig.app.json';
+      overrideJsonFile(tree2, tsCfgPath, defaultTsConfigOptions);
+      const config = getWorkspaceTargets(tree2, 'another-app');
+      config.build.options.main = 'projects\\another-app\\src\\main.ts';
+      config.build.options.polyfills = 'projects\\another-app\\src\\polyfills.ts';
+      config.build.options.tsConfig = 'projects\\another-app\\tsconfig.app.json';
+      updateWorkspaceTargets(tree2, config, 'another-app');
+
+      const tree3 = await schematicRunner.runSchematicAsync('workspace-version-9', {}, tree2.branch()).toPromise();
+      const { exclude, files } = JSON.parse(tree3.readContent(tsCfgPath));
       expect(exclude).toBeUndefined();
       expect(files).toEqual(['src/main.ts', 'src/polyfills.ts']);
     });

--- a/packages/schematics/angular/migrations/update-9/update-workspace-config_spec.ts
+++ b/packages/schematics/angular/migrations/update-9/update-workspace-config_spec.ts
@@ -13,14 +13,17 @@ import { WorkspaceTargets } from '../../utility/workspace-models';
 import { ANY_COMPONENT_STYLE_BUDGET } from './update-workspace-config';
 
 // tslint:disable-next-line: no-any
-function getWorkspaceTargets(tree: UnitTestTree): any {
-  return JSON.parse(tree.readContent(workspacePath))
-    .projects['migration-test'].architect;
+export function getWorkspaceTargets(tree: UnitTestTree, project = 'migration-test'): any {
+  return JSON.parse(tree.readContent(workspacePath)).projects[project].architect;
 }
 
-function updateWorkspaceTargets(tree: UnitTestTree, workspaceTargets: WorkspaceTargets) {
+export function updateWorkspaceTargets(
+  tree: UnitTestTree,
+  workspaceTargets: WorkspaceTargets,
+  project = 'migration-test',
+) {
   const config = JSON.parse(tree.readContent(workspacePath));
-  config.projects['migration-test'].architect = workspaceTargets;
+  config.projects[project].architect = workspaceTargets;
   tree.overwrite(workspacePath, JSON.stringify(config, undefined, 2));
 }
 

--- a/packages/schematics/angular/migrations/update-9/utils.ts
+++ b/packages/schematics/angular/migrations/update-9/utils.ts
@@ -144,3 +144,9 @@ export function isIvyEnabled(tree: Tree, tsConfigPath: string): boolean {
 
   return true;
 }
+
+// TS represents paths internally with '/' and expects paths to be in this format.
+// angular.json expects paths with '/', but doesn't enforce them.
+export function forwardSlashPath(path: string) {
+  return path.replace(/\\/g, '/');
+}

